core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -214,59 +216,82 @@ PolicySelection decideLoadBalancerProvider(
+          if (policy.equals(GRPCLB_POLICY_NAME)) {
+            return selectGrpclb(servers, backendAddrs, lbConfig.getRawConfigValue());
+          } else {
+            LoadBalancerProvider provider = registry.getProvider(policy);
+            if (provider == null) {
+              policiesTried.add(policy);
+            } else {
+              if (!policiesTried.isEmpty()) {
+                // Before returning, log all previously tried policies
+                helper.getChannelLogger().log(
+                    ChannelLogLevel.DEBUG,
+                    "{0} specified by Service Config are not available", policiesTried);
+              }
+              return new PolicySelection(provider, backendAddrs, lbConfig.getRawConfigValue());
             }
-            return new PolicySelection(provider, servers, lbConfig.getRawConfigValue());
           }
-          policiesTried.add(policy);
         }
-        throw new PolicyException(
+        if (!haveBalancerAddress) {
+          throw new PolicyException(
             "None of " + policiesTried + " specified by Service Config are available.");
+        }
+      }
+
+      if (haveBalancerAddress) {
+        // This is a special case where the existence of balancer address in the resolved address
+        // selects "grpclb" policy if the service config couldn't select a policy
+        return selectGrpclb(servers, backendAddrs, null);
       }
+      // No config nor balancer address. Use default.
       return new PolicySelection(
           getProviderOrThrow(defaultPolicy, "using default policy"), servers, null);
     }
+
+    private PolicySelection selectGrpclb(
+        List<EquivalentAddressGroup> addresses, List<EquivalentAddressGroup> backends,
+        @Nullable Map<String, Object> grpclbConfig) throws PolicyException {
+      checkArgument(addresses.size() >= backends.size(),
+          "Unexpected list sizes. addresses=%s, backends=%s", addresses, backends);
+
+      LoadBalancerProvider grpclbProvider = registry.getProvider(GRPCLB_POLICY_NAME);
+      if (grpclbProvider == null) {
+        if (backends.isEmpty()) {
+          throw new PolicyException(
+              "Received ONLY balancer addresses but grpclb runtime is missing");
+        }
+        if (!roundRobinDueToGrpclbDepMissing) {
+          roundRobinDueToGrpclbDepMissing = true;
+          String errorMsg = "Found balancer addresses but grpclb runtime is missing."
+              + " Will use round_robin. Please include grpc-grpclb in your runtime depedencies.";
+          helper.getChannelLogger().log(ChannelLogLevel.ERROR, errorMsg);
+          logger.warning(errorMsg);
+        }
+        return new PolicySelection(
+            getProviderOrThrow(
+                "round_robin", "received balancer addresses but grpclb runtime is missing"),
+            backends, null);
+      }
+      roundRobinDueToGrpclbDepMissing = false;
+      if (addresses.size() == backends.size()) {
```

Comment:
I thought this would move to gRPC-LB (and probably be a warning).

Reply:
GRPCLB is handling it correctly by failing RPCs.

Reply:
We should use the fallback policy instead of failing RPCs. Go ahead and file a bug to handle this later.